Command refactoring

cyruskarsan opened the pull request:
remove custom function command and format code with black and isort.

cyruskarsan commented:
@vringar Please review

cyruskarsan commented:
Not sure why build failed. Before pushing I ran `isort` on `command_executor.py` and it said that it formatted the file. git says everything is up to date. What am I missing?

vringar commented:
Hmm, from which folder did you run isort?
It seems the isort config we have in the root of our repository didn't get picked up

cyruskarsan commented:
@vringar I ran it from the OpenWPM root directory. Here is the command I ran (and just reran)
`isort automation/CommandSequence.py`

When I ran it the first time, it said it fixed import. Every subsequent time, isort does not print anything.

vringar commented:
Hmmmm, I'll look into this on Monday.

vringar commented on automation/Commands/command_executor.py:
```diff
@@ -25,29 +14,21 @@ def execute_command(
+        pass
 
-    elif type(command) is DumpProfCommand:
+    if type(command) is DumpProfCommand:
```
You can also remove this right?

vringar commented:
@cyruskarsan Could you run `isort . --show-config` and paste the output here.
Maybe that helps us debug the issue.

vringar commented:
I'm happy to see that isort seems to work now. You now see in the logs that two tests are broken.
Please have a look at
TestProfile::test_seed_persistance
TestCustomFunctionCommand::test_custom_function
You can run a single test locally by running:
`pytest -vv -s -k {test_name}` so e.g. `pytest -vv -s -k test_seed_persistance ` for the first test.

vringar commented:
@cyruskarsan do you want to fix these failing tests before I merge this PR or should I fix them on my branch?

cyruskarsan commented:
@vringar I'll give it a try.
When i run `pytest -vv -s -k test_seed_persistance` locally, it gives me this error which is quite strange. 
`_____ ERROR collecting test/manual_test.py ________________________________________________________________________
ImportError while importing test module '/home/cyrus/code/OpenWPM/test/manual_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../miniconda3/envs/openwpm/lib/python3.8/site-packages/_pytest/python.py:552: in _importtestmodule
    mod = import_path(self.fspath, mode=importmode)
../../miniconda3/envs/openwpm/lib/python3.8/site-packages/_pytest/pathlib.py:520: in import_path
    importlib.import_module(module_name)
../../miniconda3/envs/openwpm/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
../../miniconda3/envs/openwpm/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
test/manual_test.py:11: in <module>
    from automation import js_instrumentation as jsi
E   ModuleNotFoundError: No module named 'automation'`

It is not finding the `js_instrumentation` file which makes no sense because I checked the directory and it is there. I'm assuming there is some locally misconfiguration on my end. 

Looking at what failed in the tests, I think the tests need to get updated. The tests complain that command sequence does not have `run_custom_function` command which is true because we removed that and replaced it with the append command (I think). To resolve this issue, do we need to update the test(s)?

vringar commented:
Oh sorry. I messed up.
You need to run `python -m pytest ...` and need to be running from the `test` directory.
I have some changes to that behavior in my Data Aggregator rewrite branch so I had the wrong command in mind

vringar commented:
Hey, just checking in.
Are you still working on this PR or should I merge the changes as they are?

cyruskarsan commented:
@vringar Thanks for checking in. Is the only remaining task to correct the tests? I'm attempted to resolve the merge conflicts but it broke the program. Here is the output 
`BrowserManager       - ERROR    - BROWSER 39: Crash in driver, restarting browser manager
Traceback (most recent call last):
  File "/home/cyrus/code/OpenWPM/openwpm/BrowserManager.py", line 436, in BrowserManager
    driver, prof_folder = deploy_browser.deploy_browser(
  File "/home/cyrus/code/OpenWPM/openwpm/DeployBrowsers/deploy_browser.py", line 13, in deploy_browser
    return deploy_firefox.deploy_firefox(
  File "/home/cyrus/code/OpenWPM/openwpm/DeployBrowsers/deploy_firefox.py", line 188, in deploy_firefox
    driver.install_addon(ext_loc, temporary=True)
  File "/home/cyrus/miniconda3/envs/openwpm/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 267, in install_addon
    return self.execute("INSTALL_ADDON", payload)["value"]
  File "/home/cyrus/miniconda3/envs/openwpm/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/cyrus/miniconda3/envs/openwpm/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: No such file or directory: /home/cyrus/code/OpenWPM/openwpm/Extension/firefox/openwpm.xpi

BrowserManager       - ERROR    - BROWSER 39: Spawn unsuccessful  | Proxy Ready: False  | Profile Created: True  | Profile Tar: True  | Display: True  | Launch Attempted: True  | Browser Launched: False  | Browser Ready: False `

I committed the code but did not push. Please let me know what you think the next course of action is.

vringar commented:
You need to run `./scripts/build-extension.sh` to generate a new .xpi
Once you've done that, demo.py should run again.

vringar commented:
Hey, sorry for just merging this out of the blue.
I just got into a flow and wanted to get this wrapped up.
I hope you hadn't invested too much time into this

cyruskarsan commented:
@vringar No problem. I was going to work on this over thanksgiving but seems like you've got it :)
What next?

vringar commented:
The next think for the command refactoring would be to change the Tutorial on how to write a custom command [here](https://github.com/mozilla/OpenWPM/blob/master/docs/Using_OpenWPM.md#adding-a-new-command)

vringar commented:
Would you want to do that? Or would you rather code?

cyruskarsan commented:
@vringar I can try to write the documentation and then move onto a different project.

Here is a rough draft:
To create a new custom command, you must modify `command_sequence.py` and `browser_commands.py`
Suppose we want to make a command to take a screenshot of the page we are viewing. 
In `command_sequence.py` we would define a class, `SaveScreenshotCommand` that derives from `BaseCommand`
```python
class SaveScreenShotCommand(BaseCommand)
```

Next, in our newly derived class, we would define the `__init__`, `execute`, `___repr___` and eventually `name()` functions.
```python
def __init__(self, suffix):
        self.suffix = suffix

    def __repr__(self):
        return "SaveScreenshotCommand({})".format(self.suffix)

    def execute(
        self,
        webdriver,
        browser_params,
        manager_params,
        extension_socket,
    ):
        if self.suffix != "":
            self.suffix = "-" + self.suffix

        urlhash = md5(webdriver.current_url.encode("utf-8")).hexdigest()
        outname = os.path.join(
            manager_params["screenshot_path"],
            "%i-%s%s.png" % (self.visit_id, urlhash, self.suffix),
        )
        webdriver.save_screenshot(outname)
```

_I am not sure about this part_
Finally, we will add the command to the command sequence
`command_sequence.append(save_screenshot())`

cyruskarsan commented:
@vringar I pulled the most recent changes on the branch and noticed that when I run demo.py I get this error (as a result, I wasn't able to test my own custom command). Is this an issue on my end or with the branch?
`Traceback (most recent call last):
  File "demo.py", line 48, in <module>
    command_sequence = command_sequence.CommandSequence(
AttributeError: 'CommandSequence' object has no attribute 'CommandSequence'`

Finally, in command_sequence.py, `append_command` is duplicated. Should I fix that on my end and push to this branch?

vringar commented:
Yeah, this is issue #810. It's already fixed on master so merging master into the CommandRefactoring branch should fix this.

vringar commented:
> @vringar I can try to write the documentation and then move onto a different project.

Thank you. 
> Here is a rough draft:
> To create a new custom command, you must modify `command_sequence.py` and `browser_commands.py`

I don't think that's the case anymore. Since the whole point of the CommandRefactoring was to allow users to create a new command without having to touch OpenWPM internals.
I would instead suggest that the new command gets written in `custom_command.py` (aka a file that is not part of OpenWPM)

> Suppose we want to make a command to take a screenshot of the page we are viewing.

> In __`custom_command.py`__ we would define a class, `SaveScreenshotCommand` that derives from `BaseCommand`
> 
> ```python
> class SaveScreenShotCommand(BaseCommand)
> ```
> 
> Next, in our newly derived class, we would define the `__init__`, `execute` and `__repr__`  functions.

I figured out a way to work around the `name()` functions so we only need `execute` and `__repr__`
> ```python
> def __init__(self, suffix):
>         self.suffix = suffix
> 
>     def __repr__(self):
>         return "SaveScreenshotCommand({})".format(self.suffix)
> 
>     def execute(
>         self,
>         webdriver,
>         browser_params,
>         manager_params,
>         extension_socket,
>     ):
>         if self.suffix != "":
>             self.suffix = "-" + self.suffix
> 
>         urlhash = md5(webdriver.current_url.encode("utf-8")).hexdigest()
>         outname = os.path.join(
>             manager_params["screenshot_path"],
>             "%i-%s%s.png" % (self.visit_id, urlhash, self.suffix),
>         )
>         webdriver.save_screenshot(outname)
> ```
> 

> Finally, we will add the command to the command sequence
> `command_sequence.append_command(SaveScreenShotCommand("my_suffix"))`

This is what we added `append_command` for. So that the user could give us any object that inherits from `BaseCommand` and has an execute method.

cyruskarsan commented:
@vringar I see! That clearly went over my head! This is why I wanted to write the documentation, to make sure I understand what is going on :)

So does your revised documentation suffice or should I add more examples?

vringar commented:
I think it's sufficient for a first draft. Please open a new PR and then we can have a look if anything else needs to be changed.
